apis/v1alpha1/shared_types.go

```diff
@@ -121,6 +121,13 @@ type PortRange struct {
 }
 
 // AdminNetworkPolicyIngressPeer defines an in-cluster peer to allow traffic from.
+//
+// Note that presence of a Service object with this policy subject as its backend
+// has no impact on the behavior of the policy applied to the peer
+// trying to talk to the Service. It will work in the same way as if the
+// Service didn't exist since policy is applied after ServiceVIP (clusterIP,
+// externalIP, loadBalancerIngressVIP) is rewritten to the backendIPs.
```

Comment:
Unfortunately, this isn't quite right; ClusterIPs and ExternalIPs are always rewritten before NP enforcement, but load balancer IPs sometimes aren't; in some cases, pod-to-LoadBalancer traffic _must_ be sent to the load balancer rather than being processed locally. (This would happen either if the CloudProvider sets `hostname` rather than `ip` in the service's `LoadBalancerStatus` (like AWS), or if it sets `ip` but also uses the new [`ipMode: Proxy`](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1860-kube-proxy-IP-node-binding).)

And once a packet leaves the node and goes to an external load balancer, then all bets are off. When the packet comes back into the cluster, it's not even theoretically possible for the ANP implementation to reliably figure out its original source IP (since the LB may have transformed the connection in arbitrary ways).

(And of course, if the Service has endpoint IPs that are cluster-external, then a pod-to-LoadBalancer connection that goes to the LB would presumably then go directly to the cluster-external IPs, without coming back into the cluster at all, so there's no way the ANP implementation could enforce a `Networks` selector that blocked cluster-external IPs at that point.)

So I can think of four options:
1. Require LoadBalancer implementations to participate in ANP enforcement, so that we can guarantee that pod-to-LB connections will have the same semantics as pod-to-ClusterIP. This seems really really unlikely to happen, but I wanted to list it just for completeness.
2. Require ANP implementations to "pre-approve" pod-to-LB traffic by figuring out "if I were to rewrite this traffic myself, would it then pass ANP?", and if not, then drop it rather than forwarding it to the LB. Note that in the AWS-like case this requires using FQDN rules to recognize that traffic is headed for an LB. And then there are theoretical edge cases like "what if one of the service's endpoints is allowed but another one isn't?". So this also doesn't seem great.
3. Declare that it's undefined whether you can match post-LB traffic (IOW, a `Pods` selector may or may not catch pod-to-LB-to-pod traffic, and a `Networks` selector may or may not catch pod-to-LB-to-cluster-external traffic), but declare that you _can_ block _pre-LB_ traffic with a `Networks` selector that matches the LB IP. (This might be annoying for implementations to get right in the case where the LB traffic _doesn't_ leave the node, but it should not actually be _impossible_ for anyone.). If we were going to do this, we would probably want to say that you can block pod-to-`externalIPs` traffic too, for consistency, even though we don't run into the same "we don't know what the cloud infra is doing" problem there. (In fact, we could potentially say that this applies to `clusterIPs` too... we'd have to think that through.)
4. Declare that all pod-to-LB traffic has undefined semantics with respect to ANP, and policies may have different effects in different clusters. (This is easiest for implementors, but worst for users. And of course, this is what NP does because we hadn't noticed the problem until it was too late and there were already conflicting implementations.)

FTR, note that in the cluster-ingress case (which is even worse), I feel like there's a chance the answer will end up being "delegate policy enforcement to the Gateway rather than trying to do it in ANP". I'm not sure if that has any implications for trying to decide what to do here.

(People may also want to consider how they expect ANPs to interact with other implementation-specific egress functionality (eg, EgressIPs), and whether that implies anything about the best choice for LB behavior.)

Reply:
> Unfortunately, this isn't quite right; ClusterIPs and ExternalIPs are always rewritten before NP enforcement, but load balancer IPs sometimes aren't; in some cases, pod-to-LoadBalancer traffic _must_ be sent to the load balancer rather than being processed locally. (This would happen either if the CloudProvider sets `hostname` rather than `ip` in the service's `LoadBalancerStatus` (like AWS), or if it sets `ip` but also uses the new [`ipMode: Proxy`](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1860-kube-proxy-IP-node-binding).)
> 
> And once a packet leaves the node and goes to an external load balancer, then all bets are off. When the packet comes back into the cluster, it's not even theoretically possible for the ANP implementation to reliably figure out its original source IP (since the LB may have transformed the connection in arbitrary ways).

OH I completely forgot about the aws LB use case.. :/ great call out.

> 
> (And of course, if the Service has endpoint IPs that are cluster-external, then a pod-to-LoadBalancer connection that goes to the LB would presumably then go directly to the cluster-external IPs, without coming back into the cluster at all, so there's no way the ANP implementation could enforce a `Networks` selector that blocked cluster-external IPs at that point.)

YEA,,,,!

> 
> So I can think of four options:
> 
> 1. Require LoadBalancer implementations to participate in ANP enforcement, so that we can guarantee that pod-to-LB connections will have the same semantics as pod-to-ClusterIP. This seems really really unlikely to happen, but I wanted to list it just for completeness.
> 2. Require ANP implementations to "pre-approve" pod-to-LB traffic by figuring out "if I were to rewrite this traffic myself, would it then pass ANP?", and if not, then drop it rather than forwarding it to the LB. Note that in the AWS-like case this requires using FQDN rules to recognize that traffic is headed for an LB. And then there are theoretical edge cases like "what if one of the service's endpoints is allowed but another one isn't?". So this also doesn't seem great.

yeah except AWS case other ingress.VIP types will work well with this solution. Even with the way "Services" are implemented there is inconsistency with LBs because some are hairpinned within the cluster, some are sent out etc.. but I guess even in that case we didn't really define what is correct for implementions to follow...

> 3. Declare that it's undefined whether you can match post-LB traffic (IOW, a `Pods` selector may or may not catch pod-to-LB-to-pod traffic, and a `Networks` selector may or may not catch pod-to-LB-to-cluster-external traffic), but declare that you _can_ block _pre-LB_ traffic with a `Networks` selector that matches the LB IP. (This might be annoying for implementations to get right in the case where the LB traffic _doesn't_ leave the node, but it should not actually be _impossible_ for anyone.). 

yea... at least in ovnk.. we won't block preLB traffic..gonna be applied after the DNAT except in AWS case where we will :D 

> If we were going to do this, we would probably want to say that you can block pod-to-`externalIPs` traffic too, for consistency, even though we don't run into the same "we don't know what the cloud infra is doing" problem there. (In fact, we could potentially say that this applies to `clusterIPs` too... we'd have to think that through. This is potentially the simplest solution though, because then the rule just becomes "if the packet got DNATted, then check both the original and the new dest IP against all policies, rather than only checking the new dest IP".)

> 4. Declare that all pod-to-LB traffic has undefined semantics with respect to ANP, and policies may have different effects in different clusters. (This is easiest for implementors, but worst for users. And of course, this is what NP does because we hadn't noticed the problem until it was too late and there were already conflicting implementations.)

I like this the BEST. :D But perhaps we are then following in the footsteps of something NetPol did badly and we don't necessarily want to do that to users?

> 
> FTR, note that in the cluster-ingress case (which is even worse), I feel like there's a chance the answer will end up being "delegate policy enforcement to the Gateway rather than trying to do it in ANP". I'm not sure if that has any implications for trying to decide what to do here.
> 
> (People may also want to consider how they expect ANPs to interact with other implementation-specific egress functionality (eg, EgressIPs), and whether that implies anything about the best choice for LB behavior.)

yeah good point for cross-feature functionality user experience.

Reply:
> > 3\. Declare that it's undefined whether you can match post-LB traffic (IOW, a `Pods` selector may or may not catch pod-to-LB-to-pod traffic, and a `Networks` selector may or may not catch pod-to-LB-to-cluster-external traffic), but declare that you _can_ block _pre-LB_ traffic with a `Networks` selector that matches the LB IP. (This might be annoying for implementations to get right in the case where the LB traffic _doesn't_ leave the node, but it should not actually be _impossible_ for anyone.).
> 
> yea... at least in ovnk.. we won't block preLB traffic..gonna be applied after the DNAT except in AWS case where we will :D

ugh...

So what I was trying to say with "check both the original and the new dest IP against all policies, rather than only checking the new dest IP" was that, if you have a rule saying "Deny to 192.168.0.0/24", then instead of just checking just "`-d 192.168.0.0/24`" (iptables) or "`nw_dst=192.168.0.0/24`" (OVS), you'd also check "`-m conntrack --ctstate DNAT --ctorigdst 192.168.0.0/24`" (iptables) or "`ct_state=+dnat,ct_nw_dst=192.168.0.0/24`" (OVS), etc. You still only need to do the checks in once place, after service rewriting, you just check against both the current DNAT destination IP and the saved original destination IP.

But it looks like OVN ACL expressions don't expose OVS `ct_nw_dst`... (`ct.dnat` is the equivalent of `ct_state=+dnat`, but there's no equivalent of `ct_nw_dst` :slightly_frowning_face: )

Reply:
> > 4. Declare that all pod-to-LB traffic has undefined semantics with respect to ANP, and policies may have different effects in different clusters....
> 
> I like this the BEST. :D But perhaps we are then following in the footsteps of something NetPol did badly and we don't necessarily want to do that to users?

That's how I feel. I mean, think of the user stories that would have had to be there to get you to choose this behavior on purpose. "As an administrator, I need to be able to block certain users from accessing _X_. But if the user creates a LoadBalancer Service pointing to it, then whatever, I guess I don't care in that case."

(Admittedly, with the current default RBAC permissions, users cannot create Services that point to cluster-external IPs or to another user's pods... So... maybe we can get away with that?)

Reply:
My initial inclination is that wherever possible we should try to avoid CIDR-based matches for rich k8s constructs, because it means that every time something relocates elsewhere on the network, the user must update their network policies. The beauty of policies for Pods is that the Pods can churn as much as they like, move IPs, redeploy elsewhere, etc, and this doesn't impact the user's specification of the policy. It's just up to the implementation to learn the right information and apply the policy correctly. (And sure, we can always paper over this with controllers that react to stuff in the network and write the ANPs, but that seems like the implementation detail rather than the user-facing API; implementations should have the freedom to choose if that's the way they want it to work).

Whether this "avoid CIDR applying to services" idea means aiming for the node to do a best-effort translation and apply policy based on that, or whether to have an explicit selector that is specifically for selecting traffic towards services by the service's labels, or one of the other options... I'm not sure. 

> declare that you _can_ block _pre-LB_ traffic with a `Networks` selector that matches the LB IP. (This might be annoying for implementations to get right in the case where the LB traffic _doesn't_ leave the node, but it should not actually be _impossible_ for anyone.).

If we're talking "How to apply the policy in cases where the CNI is _not_ responsible for service translation" then I think this doable in Cilium without a huge amount of effort, but if this is intended to apply to all services including ones that the CNI is responsible for translating, then this seems to equate to building out a whole extra policy calculation + enforcement stage. So yes, a pain.